ddtrace/internal/remoteconfig/client.py

```diff
@@ -85,100 +84,148 @@ class RemoteConfigError(Exception):
-    keyval = attr.ib(type=Mapping)
-    scheme = attr.ib(type=str)
-
-
-@attr.s
-class Role(object):
-    keyids = attr.ib(type=List[str])
-    threshold = attr.ib(type=int)
-
-
-@attr.s
-class Root(object):
-    _type = attr.ib(type=str, validator=attr.validators.in_(("root",)))
-    spec_version = attr.ib(type=str)
-    consistent_snapshot = attr.ib(type=bool)
-    expires = attr.ib(type=datetime, converter=parse_isoformat)
-    keys = attr.ib(type=Mapping[str, Key])
-    roles = attr.ib(type=Mapping[str, Role])
-    version = attr.ib(type=int)
-
-
-@attr.s
-class SignedRoot(object):
-    signatures = attr.ib(type=List[Signature])
-    signed = attr.ib(type=Root)
-
-
-@attr.s
-class TargetDesc(object):
-    length = attr.ib(type=int)
-    hashes = attr.ib(type=Mapping[str, str])
-    custom = attr.ib(type=Mapping[str, Any])
-
-
-@attr.s
-class Targets(object):
-    _type = attr.ib(type=str, validator=attr.validators.in_(("targets",)))
-    custom = attr.ib(type=Mapping[str, Any])
-    expires = attr.ib(type=datetime, converter=parse_isoformat)
-    spec_version = attr.ib(type=str, validator=attr.validators.in_(("1.0", "1.0.0")))
-    targets = attr.ib(type=Mapping[str, TargetDesc])
-    version = attr.ib(type=int)
-
-
-@attr.s
-class SignedTargets(object):
-    signatures = attr.ib(type=List[Signature])
-    signed = attr.ib(type=Targets)
-
-
-@attr.s
-class TargetFile(object):
-    path = attr.ib(type=str)
-    raw = attr.ib(type=str)
-
-
-@attr.s
-class AgentPayload(object):
-    roots = attr.ib(type=List[SignedRoot], default=None)
-    targets = attr.ib(type=SignedTargets, default=None)
-    target_files = attr.ib(type=List[TargetFile], default=[])
-    client_configs = attr.ib(type=Set[str], default=set())
+    id: str
+    product_name: str
+    sha256_hash: Optional[str]
+    length: Optional[int]
+    tuf_version: Optional[int]
+    apply_state: Optional[int] = dataclasses.field(default=1, compare=False)
+    apply_error: Optional[str] = dataclasses.field(default=None, compare=False)
+
+
+@dataclasses.dataclass
+class Signature:
+    keyid: str
+    sig: str
+
+
+@dataclasses.dataclass
+class Key:
+    keytype: str
+    keyid_hash_algorithms: List[str]
+    keyval: Mapping
+    scheme: str
+
+
+@dataclasses.dataclass
+class Role:
+    keyids: List[str]
+    threshold: int
+
+
+@dataclasses.dataclass
+class Root:
+    _type: str
+    spec_version: str
+    consistent_snapshot: bool
+    expires: datetime
+    keys: Mapping[str, Key]
+    roles: Mapping[str, Role]
+    version: int
+
+    def __post_init__(self):
+        if self._type != "root":
+            raise ValueError("Root: invalid root type")
+        if isinstance(self.expires, str):
+            self.expires = parse_isoformat(self.expires)
+        for k, v in self.keys.items():
+            if isinstance(v, dict):
+                self.keys[k] = Key(**v)
+        for k, v in self.roles.items():
+            if isinstance(v, dict):
+                self.roles[k] = Role(**v)
```

Comment:
nit: In the `__post_init__()`s, are we OK with leaving the values that don't conform to the specified type as is?

Reply:
We have definitely less dynamic checks than with attr, but those classes are only used in that file, so we can be sure it will be under control. Here failures to conform can only come from an unexpected RC payload that will be rejected.